### PR TITLE
gh-114099: Add test exclusions to support running the test suite on iOS

### PR DESCRIPTION
**Pull request by freakboy3742**
Part of the PEP 730 work to add iOS support.

Adds test annotations required to run the CPython test suite on iOS.

The majority of these changes involve:
* Annotating tests that use subprocess, but are skipped on Emscripten/WASI for other reasons
* Including iOS/tvOS/watchOS under the same umbrella as macOS/darwin checks

`is_apple` and `is_apple_mobile` test helpers have been added to identify *any* Apple platform, and "any Apple platform except macOS", respectively.

Refs #114099.

* Issue: gh-114099

### CODE REVIEW

**Review by erlend-aasland — APPROVED** [review]
Thanks, looks good to me!

File: Lib/test/support/__init__.py
Comment (mhsmith) [inline Lib/test/support/__init__.py]:
```suggestion
    and not (is_emscripten or is_wasi or is_apple_mobile)
```

File: Lib/test/support/__init__.py
Comment (mhsmith) [inline Lib/test/support/__init__.py]:
```suggestion
has_subprocess_support = not (is_emscripten or is_wasi or is_apple_mobile)
```

File: Lib/test/test_asyncio/test_events.py
Comment (mhsmith) [inline Lib/test/test_asyncio/test_events.py]: Applying the decorator to the mixin class would cause its subclasses to skip all of their tests, not just the subprocess ones. I've got a possible alternative approach on [this branch](https://github.com/python/cpython/commit/460078deda6a9adaa13d4b4874cf4f0f03b717a8#diff-b125df0ba39acde9a49ff4ce6a554310cab3daf258a2e77c219fce5a8b17c146).

File: Lib/test/test_asyncio/test_unix_events.py
Comment (mhsmith) [inline Lib/test/test_asyncio/test_unix_events.py]:
```suggestion
@support.requires_fork()
```

File: Lib/test/test_pty.py
Comment (mhsmith) [inline Lib/test/test_pty.py]:
```suggestion
# Skip these tests if termios is not available
```

File: Lib/test/test_venv.py
Comment (mhsmith) [inline Lib/test/test_venv.py]: Suggest combining these two blocks into one, as in test_pty.

File: Lib/test/support/__init__.py
Comment (freakboy3742) [reply Lib/test/support/__init__.py]: I'll defer to the opinion of the core team, but for my money, the "one per line" format is more readable.

File: Lib/test/support/__init__.py
Comment (freakboy3742) [reply Lib/test/support/__init__.py]: As above; for my money, the "one per line" format is more readable.

File: Lib/test/test_asyncio/test_events.py
Comment (freakboy3742) [reply Lib/test/test_asyncio/test_events.py]: Agreed that this is possibly over-enthusiastic in what it is skipping; my inclination is to be explicit on each test that should be skipped, rather than programmatically excluding every test, on the off chance that a test is ever added to the mixin that *doesn't* explicitly require subprocess.

File: Lib/test/support/__init__.py
Comment (mhsmith) [reply Lib/test/support/__init__.py]: Agreed on the one item per line – my suggestion was more about switching from `and` to `or`.

Comment (freakboy3742) [reply Lib/test/support/__init__.py]: Makes sense - I've made that simplification.

**Review by mhsmith — APPROVED** [review]

**Comment by mhsmith** [comment]
I'm not a project member, so I can't resolve my conversations, but they've all now been addressed.

File: Lib/test/test_asyncio/test_unix_events.py
Comment (willingc) [inline Lib/test/test_asyncio/test_unix_events.py:1877]: Nice. Much cleaner.

File: Lib/test/test_code_module.py
Comment (willingc) [inline Lib/test/test_code_module.py:163]: Does this work if the `import site` does not exist in the file?

**Review by willingc — APPROVED** [review]
Looks good. Ran without issue on Sonoma 14.2.1. One small question/nit re: an import.
:ship:

File: Lib/test/test_code_module.py
Comment (freakboy3742) [reply Lib/test/test_code_module.py:163]: It does - the `import site` was a leftover from a previous attempt to fix the problem that was ultimately resolved with the `no_site`-based skip. I've removed the import.

**Comment by erlend-aasland** [comment]
Are you ready to land this, @freakboy3742?

**Comment by freakboy3742** [comment]
@erlend-aasland I'm ready if everyone else is :-)

**Comment by erlend-aasland** [comment]
AFAICS, @willingc's remarks have been addressed; the rest are thumbs up. Let's go :)

File: Lib/test/test_fcntl.py
Comment (encukou) [inline Lib/test/test_fcntl.py:11]: You seem to do this change often. In the future, could you add a trailing comma, to make future diffs nicer?
(Or simply add another line if the current one gets too long:
```
from test.support import verbose, cpython_only, get_pagesize
from test.support import is_apple, requires_subprocess
```
)

Comment (freakboy3742) [reply Lib/test/test_fcntl.py:11]: @encukou Sure, no problem. 

One of the issues I've hit is that the "preferred" style for imports varies wildly between files. I appreciate there's 30 years of historical baggage in the existing code - but is there a reason either of those two approaches is preferred over isort/black style syntax:
```
from test.support import (
    cpython_only, 
    get_pagesize, 
    is_apple, 
    requires_subprocess, 
    verbose,
)
```
Running black over every file I touch would make for unreadable patches; but if little black touches when the line is already being altered would be looked on favourably, then that's an easy tweak to make.

Comment (encukou) [reply Lib/test/test_fcntl.py:11]: That works too -- there's a [trailing comma](https://peps.python.org/pep-0008/#when-to-use-trailing-commas), so future diffs won't need to add it.

(The 30 years of baggage also means there is no shared preference for stuff that isn't in PEP 8. I'm partial to repeating `from <name>` because it's searchable with `grep` -- you don't need a code analyzer -- but that's just personal opinion.)

> if little black touches when the line is already being altered would be looked on favourably

Sure! As long as it helps readability :)

**Pull request merged**
